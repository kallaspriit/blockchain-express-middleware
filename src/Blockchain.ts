import Axios from "axios";
import * as querystring from "querystring";
import { dummyLogger, ILogger } from "./abstract-logger";
import { Invoice } from "./index";

/**
 * Base configuration that has reasonable defaults and do no require to be redefined by the user.
 */
export interface IBlockchainBaseConfig {
  apiBaseUrl: string;
  gapLimit?: number;
}

/**
 * Configuration required to be provided by the user.
 */
export interface IBlockchainRequiredConfig {
  apiKey: string;
  xPub: string;
}

/**
 * Configuration that the user provides, includes mandatory user configuration and optional base configuration.
 */
export type IBlockchainUserConfig = Partial<IBlockchainBaseConfig> & IBlockchainRequiredConfig;

/**
 * Combined configuration including base and user provided configurations.
 */
export type IBlockchainConfig = IBlockchainBaseConfig & IBlockchainRequiredConfig;

/**
 * Parameters for the generate receiving address endpoint.
 */
export interface IGenerateReceivingAddressParameters {
  xpub: string;
  callback: string;
  key: string;
  gap_limit?: number;
}

/**
 * Response for the generate receiving address endpoint.
 */
export interface IReceivingAddress {
  address: string;
  index: number;
  callback: string;
}

/**
 * Parameters for creating an invoice.
 */
export interface ICreateInvoiceInfo {
  dueAmount: number;
  message: string;
  secret: string;
  callbackUrl: string;
}

/**
 * Default base configuration.
 */
export const defaultBaseConfig: IBlockchainBaseConfig = {
  apiBaseUrl: "https://api.blockchain.info/v2/receive",
};

/**
 * Provides API for receiving payments through blockchain.info service.
 *
 * To get the extended public key (xPub), copy your mnemonic words to http://bip32.org/ "Passphrase" field, wait for it
 * to generate the extended key and use the value of "Derived Public Key" as the "xPub" parameter.
 *
 * See https://blockchain.info/api/api_receive for API documentation.
 */
export default class Api {
  private readonly config: IBlockchainConfig;

  /**
   * Constructor.
   *
   * Accepts configuration and optional logger to use.
   *
   * @param userConfig User configuration (can override base configuration as well)
   * @param log Logger to use (defaults to console, but you can use bunyan etc)
   */
  public constructor(userConfig: IBlockchainUserConfig, private readonly log: ILogger = dummyLogger) {
    this.config = {
      ...defaultBaseConfig,
      ...userConfig,
    };
  }

  /**
   * Generates a new receiving address.
   *
   * @param callbackUrl URL to call on new transactions and confirmation count changes
   */
  public async generateReceivingAddress(callbackUrl: string): Promise<IReceivingAddress> {
    const { apiBaseUrl, xPub, apiKey, gapLimit } = this.config;
    const parameters: IGenerateReceivingAddressParameters = {
      xpub: xPub,
      callback: callbackUrl,
      key: apiKey,
    };

    // only add the gap limit parameters if it has been set
    if (gapLimit !== undefined) {
      parameters.gap_limit = gapLimit;
    }

    // build the request url
    const url = `${apiBaseUrl}?${querystring.stringify(parameters)}`;

    this.log.info(
      {
        parameters,
        url,
      },
      "generating receiving address",
    );

    // attempt to generate the receiving address (throws for non 2xx response)
    try {
      const response = await Axios.get<IReceivingAddress>(url);

      return response.data;
    } catch (error) {
      // log failure and rethrow the error
      this.log.error(
        {
          message: error.message,
          status: error.response.status,
          statusText: error.response.statusText,
          data: error.response.data,
          parameters,
          url,
        },
        "generating receiving address failed",
      );

      throw error;
    }
  }

  /**
   * Creates a new invoice.
   *
   * First generates the receiving address and then the invoice.
   *
   * @param info Invoice info
   */
  public async createInvoice({ dueAmount, message, secret, callbackUrl }: ICreateInvoiceInfo): Promise<Invoice> {
    // build invoice signature to later verify that the handle payment request is valid
    const signature = Invoice.getInvoiceSignature(
      {
        dueAmount,
        message,
      },
      secret,
    );

    // build the callback url, containing invoice info signed with the secret
    const decoratedCallbackUrl = `${callbackUrl}?${querystring.stringify({ signature })}`;

    // generate next receiving address
    const receivingAddress = await this.generateReceivingAddress(decoratedCallbackUrl);

    // create a new invoice
    const invoice = new Invoice({
      dueAmount,
      message,
      address: receivingAddress.address,
    });

    return invoice;
  }
}
