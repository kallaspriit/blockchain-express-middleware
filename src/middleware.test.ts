import * as bodyParser from "body-parser";
import * as express from "express";
import * as HttpStatus from "http-status-codes";
import * as querystring from "querystring";
import * as supertest from "supertest";
import blockchainMiddleware, { IInvoice, ILogger, Invoice } from "./";
import { processInvoiceForSnapshot } from "./Blockchain.test";

const SECRET = "zzz";
const RECEIVING_ADDRESS = "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq";

// invoices "database" emulated with a simple array (store the data only)
const invoiceDatabase: IInvoice[] = [];

let server: supertest.SuperTest<supertest.Test>;

// tslint:disable:no-magic-numbers
describe("middleware", () => {
  beforeEach(() => {
    // create app
    const app = express();

    // parse application/x-www-form-urlencoded and  application/json
    app.use(bodyParser.urlencoded({ extended: false }));
    app.use(bodyParser.json());

    // use the blockchain middleware
    app.use(
      "/payment",
      blockchainMiddleware({
        secret: SECRET,
        requiredConfirmations: 3,
        saveInvoice,
        loadInvoice,
      }),
    );

    server = supertest(app);
  });

  afterEach(() => {
    // "clear" the database
    while (invoiceDatabase.length > 0) {
      invoiceDatabase.pop();
    }
  });

  it("should provide qr code rendering", async () => {
    // build qr code url
    const qrCodeParameters = {
      address: RECEIVING_ADDRESS,
      amount: 1,
      message: "Test",
    };

    const response = await server.get(`/payment/qr?${querystring.stringify(qrCodeParameters)}`);

    expect(response.type).toEqual("image/png");
    expect(response.body).toMatchSnapshot();
  });

  // TODO: test invalid qr code parameters

  it("should handle valid payment updates", async () => {
    const invoice = new Invoice({
      dueAmount: 10,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoiceDatabase.push(invoice.toJSON());

    // no confirmations
    let parameters = {
      signature: invoice.getSignature(SECRET),
      address: RECEIVING_ADDRESS,
      transaction_hash: "xxx",
      value: 10,
      confirmations: 0,
    };

    const response1 = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response1.status).toEqual(HttpStatus.OK);
    expect(response1.text).toMatchSnapshot();
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();

    // now has enough confirmations
    parameters = {
      signature: invoice.getSignature(SECRET),
      address: RECEIVING_ADDRESS,
      transaction_hash: "xxx",
      value: 10,
      confirmations: 5,
    };

    const response2 = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response2.status).toEqual(HttpStatus.OK);
    expect(response2.text).toEqual("*ok*");
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();
  });

  it("should handle invalid invoice address", async () => {
    expect(invoiceDatabase.length).toEqual(0);

    const invoice = new Invoice({
      dueAmount: 10,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoiceDatabase.push(invoice.toJSON());

    // has invalid address
    const parameters = {
      signature: invoice.getSignature(SECRET),
      address: "foobar",
      transaction_hash: "xxx",
      value: 10,
      confirmations: 0,
    };

    const response = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response.status).toEqual(HttpStatus.OK);
    expect(response.text).toEqual("*ok*");
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();
  });

  it("should ignore invalid update", async () => {
    const invoice = new Invoice({
      dueAmount: 10,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoiceDatabase.push(invoice.toJSON());

    // signature is made with an invalid secret
    const parameters = {
      signature: invoice.getSignature("foobar"),
      address: RECEIVING_ADDRESS,
      transaction_hash: "xxx",
      value: 10,
      confirmations: 0,
    };

    const response = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response.status).toEqual(HttpStatus.BAD_REQUEST);
    expect(response.text).toMatchSnapshot();
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();
  });

  it("should fail if update tries to change amount", async () => {
    const invoice = new Invoice({
      dueAmount: 10,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoiceDatabase.push(invoice.toJSON());

    // has value of 10
    let parameters = {
      signature: invoice.getSignature(SECRET),
      address: RECEIVING_ADDRESS,
      transaction_hash: "xxx",
      value: 10,
      confirmations: 0,
    };

    const response1 = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response1.status).toEqual(HttpStatus.OK);
    expect(response1.text).toMatchSnapshot();
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();

    // has new value of 100
    parameters = {
      signature: invoice.getSignature(SECRET),
      address: RECEIVING_ADDRESS,
      transaction_hash: "xxx",
      value: 100,
      confirmations: 5,
    };

    const response2 = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response2.status).toEqual(HttpStatus.BAD_REQUEST);
    expect(response2.text).toMatchSnapshot();
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();
  });

  it("should ignore updates once already complete", async () => {
    const invoice = new Invoice({
      dueAmount: 10,
      message: "Test",
      address: RECEIVING_ADDRESS,
    });

    invoiceDatabase.push(invoice.toJSON());

    // has already enough confirmations
    let parameters = {
      signature: invoice.getSignature(SECRET),
      address: RECEIVING_ADDRESS,
      transaction_hash: "xxx",
      value: 10,
      confirmations: 5,
    };

    const response1 = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response1.status).toEqual(HttpStatus.OK);
    expect(response1.text).toEqual("*ok*");
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();

    // has even more confirmations
    parameters = {
      signature: invoice.getSignature(SECRET),
      address: RECEIVING_ADDRESS,
      transaction_hash: "xxx",
      value: 10,
      confirmations: 10,
    };

    const response2 = await server.get(`/payment/handle-payment?${querystring.stringify(parameters)}`);

    expect(response2.status).toEqual(HttpStatus.OK);
    expect(response2.text).toEqual("*ok*");
    expect(processInvoicesDatabaseForSnapshot(invoiceDatabase)).toMatchSnapshot();
  });

  it("should accept custom logger", async () => {
    const mockLogger: ILogger = {
      trace: jest.fn(),
      debug: jest.fn(),
      info: jest.fn(),
      warn: jest.fn(),
      error: jest.fn(),
    };

    blockchainMiddleware({
      secret: SECRET,
      requiredConfirmations: 3,
      saveInvoice,
      loadInvoice,
      log: mockLogger,
    });
  });
});

async function saveInvoice(invoice: Invoice): Promise<void> {
  // find invoice by address
  const index = invoiceDatabase.findIndex(item => item.address === invoice.address);

  // update existing invoice if exists, otherwise add a new one
  if (index !== -1) {
    invoiceDatabase[index] = invoice.toJSON();
  } else {
    invoiceDatabase.push(invoice.toJSON());
  }
}

async function loadInvoice(address: string): Promise<Invoice | undefined> {
  // search for invoice by address
  const invoiceInfo = invoiceDatabase.find(item => item.address === address);

  // return undefined if not found
  if (!invoiceInfo) {
    return undefined;
  }

  // de-serialize the invoice
  return new Invoice(invoiceInfo);
}

function processInvoicesDatabaseForSnapshot(invoices: IInvoice[]): IInvoice[] {
  invoices.forEach(processInvoiceForSnapshot);

  return invoices;
}
