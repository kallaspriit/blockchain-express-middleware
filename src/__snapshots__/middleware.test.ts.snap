// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain should accept custom logger 1`] = `
Array [
  Array [
    Object {
      "parameters": Object {
        "callback": "https://example.com?signature=4b506b7c3a37df75af9ac0a54e810ce3397fbe2255d9d820c041f9bec3a3516cfc39bce02570dfed230d81f5f37f0ba5bf886275c69d1e8b31399df45a250dd2",
        "key": "xxx",
        "xpub": "yyy",
      },
      "url": "https://api.blockchain.info/v2/receive?xpub=yyy&callback=https%3A%2F%2Fexample.com%3Fsignature%3D4b506b7c3a37df75af9ac0a54e810ce3397fbe2255d9d820c041f9bec3a3516cfc39bce02570dfed230d81f5f37f0ba5bf886275c69d1e8b31399df45a250dd2&key=xxx",
    },
    "generating receiving address",
  ],
]
`;

exports[`Blockchain should create a new invoice with receiving address 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "createdDate": 2018-04-19T13:48:05.316Z,
  "dueAmount": 1,
  "message": "Test invoice",
  "paymentState": "PENDING",
  "stateTransitions": Array [],
  "transactions": Array [],
  "updatedDate": 2018-04-20T13:48:05.316Z,
}
`;

exports[`Blockchain should generate a new receiving address 1`] = `
Object {
  "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
  "callback": "https://example.com",
  "index": 0,
}
`;

exports[`Blockchain should throw error when generating receving address and getting a non 2xx response 1`] = `[Error: Request failed with status code 400]`;

exports[`middleware should fail if update tries to change amount 1`] = `"pending"`;

exports[`middleware should fail if update tries to change amount 2`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "WAITING_FOR_CONFIRMATION",
    "stateTransitions": Array [
      Object {
        "date": 2018-04-21T13:48:05.316Z,
        "newState": "WAITING_FOR_CONFIRMATION",
        "previousState": "PENDING",
      },
    ],
    "transactions": Array [
      Object {
        "amount": 10,
        "confirmations": 0,
        "createdDate": 2018-04-19T13:48:05.316Z,
        "hash": "xxx",
        "updatedDate": 2018-04-20T13:48:05.316Z,
      },
    ],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should fail if update tries to change amount 3`] = `"got bad transaction"`;

exports[`middleware should fail if update tries to change amount 4`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "WAITING_FOR_CONFIRMATION",
    "stateTransitions": Array [
      Object {
        "date": 2018-04-21T13:48:05.316Z,
        "newState": "WAITING_FOR_CONFIRMATION",
        "previousState": "PENDING",
      },
    ],
    "transactions": Array [
      Object {
        "amount": 10,
        "confirmations": 0,
        "createdDate": 2018-04-19T13:48:05.316Z,
        "hash": "xxx",
        "updatedDate": 2018-04-20T13:48:05.316Z,
      },
    ],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should handle invalid invoice address 1`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "PENDING",
    "stateTransitions": Array [],
    "transactions": Array [],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should handle valid payment updates 1`] = `"pending"`;

exports[`middleware should handle valid payment updates 2`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "WAITING_FOR_CONFIRMATION",
    "stateTransitions": Array [
      Object {
        "date": 2018-04-21T13:48:05.316Z,
        "newState": "WAITING_FOR_CONFIRMATION",
        "previousState": "PENDING",
      },
    ],
    "transactions": Array [
      Object {
        "amount": 10,
        "confirmations": 0,
        "createdDate": 2018-04-19T13:48:05.316Z,
        "hash": "xxx",
        "updatedDate": 2018-04-20T13:48:05.316Z,
      },
    ],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should handle valid payment updates 3`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "CONFIRMED",
    "stateTransitions": Array [
      Object {
        "date": 2018-04-21T13:48:05.316Z,
        "newState": "WAITING_FOR_CONFIRMATION",
        "previousState": "PENDING",
      },
      Object {
        "date": 2018-04-21T13:48:05.316Z,
        "newState": "CONFIRMED",
        "previousState": "WAITING_FOR_CONFIRMATION",
      },
    ],
    "transactions": Array [
      Object {
        "amount": 10,
        "confirmations": 5,
        "createdDate": 2018-04-19T13:48:05.316Z,
        "hash": "xxx",
        "updatedDate": 2018-04-20T13:48:05.316Z,
      },
    ],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should ignore invalid update 1`] = `"got invalid payment status update"`;

exports[`middleware should ignore invalid update 2`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "PENDING",
    "stateTransitions": Array [],
    "transactions": Array [],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should ignore updates once already complete 1`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "CONFIRMED",
    "stateTransitions": Array [
      Object {
        "date": 2018-04-21T13:48:05.316Z,
        "newState": "CONFIRMED",
        "previousState": "PENDING",
      },
    ],
    "transactions": Array [
      Object {
        "amount": 10,
        "confirmations": 5,
        "createdDate": 2018-04-19T13:48:05.316Z,
        "hash": "xxx",
        "updatedDate": 2018-04-20T13:48:05.316Z,
      },
    ],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should ignore updates once already complete 2`] = `
Array [
  Object {
    "address": "2FupTEd3PDF7HVxNrzNqQGGoWZA4rqiphq",
    "createdDate": 2018-04-19T13:48:05.316Z,
    "dueAmount": 10,
    "message": "Test",
    "paymentState": "CONFIRMED",
    "stateTransitions": Array [
      Object {
        "date": 2018-04-21T13:48:05.316Z,
        "newState": "CONFIRMED",
        "previousState": "PENDING",
      },
    ],
    "transactions": Array [
      Object {
        "amount": 10,
        "confirmations": 5,
        "createdDate": 2018-04-19T13:48:05.316Z,
        "hash": "xxx",
        "updatedDate": 2018-04-20T13:48:05.316Z,
      },
    ],
    "updatedDate": 2018-04-20T13:48:05.316Z,
  },
]
`;

exports[`middleware should provide qr code rendering 1`] = `
Object {
  "data": Array [
    137,
    80,
    78,
    71,
    13,
    10,
    26,
    10,
    0,
    0,
    0,
    13,
    73,
    72,
    68,
    82,
    0,
    0,
    0,
    180,
    0,
    0,
    0,
    180,
    8,
    0,
    0,
    0,
    0,
    24,
    166,
    89,
    238,
    0,
    0,
    2,
    94,
    73,
    68,
    65,
    84,
    120,
    218,
    237,
    220,
    193,
    110,
    2,
    49,
    12,
    69,
    209,
    249,
    255,
    159,
    110,
    87,
    149,
    80,
    133,
    243,
    174,
    157,
    12,
    2,
    113,
    103,
    67,
    53,
    20,
    114,
    82,
    53,
    181,
    99,
    135,
    94,
    63,
    31,
    120,
    93,
    162,
    69,
    139,
    22,
    45,
    90,
    180,
    104,
    209,
    162,
    69,
    191,
    41,
    250,
    90,
    92,
    127,
    207,
    63,
    126,
    239,
    227,
    107,
    210,
    253,
    106,
    12,
    50,
    174,
    104,
    209,
    59,
    232,
    167,
    191,
    252,
    112,
    240,
    255,
    247,
    171,
    73,
    84,
    143,
    104,
    92,
    209,
    162,
    7,
    232,
    213,
    2,
    172,
    22,
    95,
    181,
    120,
    86,
    216,
    238,
    184,
    162,
    69,
    159,
    70,
    147,
    192,
    147,
    38,
    75,
    2,
    146,
    104,
    209,
    175,
    68,
    87,
    127,
    248,
    171,
    215,
    146,
    197,
    40,
    90,
    244,
    221,
    232,
    213,
    100,
    158,
    13,
    152,
    54,
    12,
    85,
    80,
    26,
    37,
    106,
    162,
    69,
    55,
    209,
    52,
    128,
    156,
    126,
    188,
    125,
    55,
    46,
    250,
    43,
    209,
    177,
    248,
    7,
    18,
    161,
    106,
    131,
    75,
    10,
    57,
    183,
    84,
    77,
    69,
    127,
    53,
    154,
    36,
    65,
    85,
    225,
    176,
    66,
    63,
    123,
    111,
    18,
    156,
    198,
    17,
    81,
    180,
    104,
    176,
    208,
    82,
    82,
    147,
    10,
    52,
    233,
    235,
    132,
    23,
    45,
    122,
    138,
    78,
    137,
    83,
    90,
    68,
    211,
    64,
    147,
    54,
    205,
    162,
    69,
    79,
    209,
    49,
    1,
    135,
    27,
    91,
    146,
    208,
    167,
    231,
    68,
    139,
    62,
    129,
    238,
    32,
    82,
    128,
    32,
    197,
    244,
    213,
    61,
    209,
    162,
    79,
    160,
    19,
    106,
    133,
    72,
    193,
    131,
    140,
    67,
    54,
    188,
    162,
    69,
    79,
    209,
    169,
    41,
    68,
    130,
    72,
    218,
    204,
    166,
    201,
    136,
    22,
    189,
    139,
    238,
    4,
    137,
    78,
    81,
    60,
    53,
    67,
    233,
    65,
    44,
    209,
    162,
    167,
    185,
    199,
    42,
    8,
    84,
    131,
    118,
    22,
    46,
    77,
    152,
    90,
    221,
    45,
    209,
    162,
    67,
    163,
    40,
    5,
    128,
    78,
    163,
    159,
    108,
    8,
    104,
    210,
    36,
    90,
    244,
    78,
    194,
    68,
    27,
    64,
    157,
    131,
    83,
    147,
    205,
    172,
    104,
    209,
    39,
    208,
    171,
    70,
    17,
    109,
    114,
    86,
    11,
    151,
    252,
    144,
    70,
    39,
    107,
    68,
    139,
    222,
    40,
    162,
    76,
    14,
    136,
    167,
    73,
    165,
    73,
    138,
    22,
    189,
    115,
    232,
    123,
    149,
    72,
    165,
    69,
    149,
    146,
    44,
    82,
    192,
    108,
    5,
    23,
    209,
    162,
    33,
    154,
    22,
    96,
    232,
    98,
    36,
    27,
    102,
    209,
    162,
    79,
    163,
    83,
    242,
    211,
    105,
    20,
    209,
    34,
    123,
    167,
    16,
    41,
    90,
    116,
    183,
    150,
    215,
    109,
    200,
    167,
    98,
    205,
    106,
    243,
    218,
    41,
    74,
    138,
    22,
    61,
    65,
    119,
    19,
    118,
    90,
    188,
    76,
    27,
    3,
    218,
    248,
    20,
    45,
    186,
    139,
    38,
    13,
    126,
    90,
    8,
    95,
    37,
    83,
    177,
    112,
    62,
    237,
    4,
    136,
    22,
    29,
    192,
    116,
    99,
    64,
    112,
    41,
    216,
    144,
    67,
    44,
    162,
    69,
    159,
    200,
    61,
    232,
    6,
    54,
    5,
    149,
    180,
    72,
    183,
    119,
    46,
    162,
    69,
    15,
    63,
    24,
    76,
    27,
    59,
    100,
    34,
    157,
    195,
    44,
    162,
    69,
    79,
    208,
    244,
    80,
    96,
    42,
    70,
    210,
    224,
    145,
    14,
    185,
    136,
    22,
    125,
    2,
    157,
    146,
    242,
    171,
    121,
    77,
    131,
    78,
    187,
    19,
    32,
    90,
    244,
    193,
    127,
    242,
    183,
    106,
    202,
    167,
    194,
    229,
    145,
    44,
    79,
    180,
    232,
    33,
    58,
    190,
    33,
    56,
    144,
    149,
    10,
    235,
    91,
    135,
    84,
    68,
    139,
    30,
    162,
    201,
    7,
    28,
    104,
    32,
    34,
    80,
    209,
    162,
    119,
    209,
    221,
    5,
    214,
    57,
    48,
    59,
    89,
    196,
    162,
    69,
    159,
    168,
    79,
    211,
    195,
    86,
    171,
    3,
    38,
    157,
    123,
    228,
    80,
    174,
    104,
    209,
    93,
    244,
    39,
    92,
    162,
    69,
    139,
    22,
    45,
    90,
    180,
    104,
    209,
    162,
    69,
    191,
    201,
    245,
    11,
    5,
    115,
    227,
    129,
    113,
    208,
    243,
    118,
    0,
    0,
    0,
    0,
    73,
    69,
    78,
    68,
    174,
    66,
    96,
    130,
  ],
  "type": "Buffer",
}
`;
